effects/game-of-life/game-of-life.c

```diff
@@ -0,0 +1,427 @@
+  bplptr[2] = CopMove32(cp, bplpt[2], prev_states[2]->planes[0]);
+  bplptr[3] = CopMove32(cp, bplpt[3], prev_states[3]->planes[0]);
+  for (i = 1; i <= DISP_HEIGHT; i += 2)
+  {
+    // vertical pixel doubling
+    CopMove16(cp, bpl1mod, -prev_states[0]->bytesPerRow);
+    CopMove16(cp, bpl2mod, -prev_states[0]->bytesPerRow);
+    CopWaitSafe(cp, Y(i), 0);
+    CopMove16(cp, bpl1mod, 0);
+    CopMove16(cp, bpl2mod, 0);
+    CopWaitSafe(cp, Y(i+1), 0);
+  }
+  CopEnd(cp);
+}
+
+static void UpdateBitplanePointers(void)
+{
+  BitmapT* cur;
+  u_short i;
+  for (i = 0; i < DISP_DEPTH; i++)
+  {
+    // update bitplane order: (states_head + i + 1) % DISP_DEPTH iterates from
+    // the oldest to newest game state, so 0th bitplane displays the oldest state
+    // and (DISP_DEPTH-1)'th bitplane displays the newest state
+    cur = prev_states[(states_head + i + 1) % DISP_DEPTH];
+    CopInsSet32(bplptr[i], cur->planes[0]);
+  }
+}
+
+static void GameOfLife(void)
+{
+  Area2D clearArea = { laser_beam_pos, 0, 8, 256 };
+  ClearIRQ(INTF_BLIT);
+  switch (phase)
+  {
+    // sum horizontally adjacent pixels - produces two results (sum bits): lo and hi
+    case 0: BlitAdjacentHorizontal(current_board, lo, minterms_table[0]); break;
+    case 1: BlitAdjacentHorizontal(current_board, hi, minterms_table[1]); break;
+
+    // result based on the number of set bits in lo
+    // set bits -> result
+    // 0        -> 0
+    // 1        -> 1
+    // 2        -> 1
+    // 3        -> 0
+    case 2: BlitAdjacentVertical(lo, x0, minterms_table[2]); break;
+    // set bits -> result
+    // 0        -> 1
+    // 1        -> 0
+    // 2        -> 1
+    // 3        -> 0
+    case 3: BlitAdjacentVertical(lo, x1, minterms_table[3]); break;
+
+    // result based on the number of set bits in hi
+    // set bits -> result
+    // 0 -> 1
+    // 1 -> 0
+    // 2 -> 1
+    // 3 -> 0
+    case 4: BlitAdjacentVertical(hi, x2, minterms_table[4]); break;
+    // set bits -> result
+    // 0 -> 1
+    // 1 -> 0
+    // 2 -> 0
+    // 3 -> 1
+    case 5: BlitAdjacentVertical(hi, x3, minterms_table[5]); break;
+
+    // black magic happens here - combines previous results and initial game state into new state
+    case 6: BlitFunc(x2, current_board, x3, x5, minterms_table[6]); break;
+    case 7: BlitFunc(x1, x5, x3, x6, minterms_table[7]); break;
+    case 8: BlitFunc(x2, x0, x6, current_board, minterms_table[8]); break;
+
+    // clear a column of cells where the beam is
+    case 9: BitmapClearArea(current_board, &clearArea); break;  // bug: this should be performed on different area
+    case 10: BitmapClearArea(prev_states[(states_head + 1) % DISP_DEPTH], &clearArea); break;
+    case 11: BitmapClearArea(prev_states[(states_head + 2) % DISP_DEPTH], &clearArea); break;
+    case 12: BitmapClearArea(prev_states[(states_head + 3) % DISP_DEPTH], &clearArea); break;
+
+    // hack - avoid graphical artifacts
+    case 13: WaitVBlank();
```

Comment:
Ouch... that hurts. Waiting for VBL in interrupt handler is a bad practice :/

Reply:
This is a hack anyway that should be resolved when we move to double buffering.